In [1]:
#Initial imports 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder, MinMaxScaler
from sklearn.metrics import accuracy_score, f1_score, recall_score, roc_auc_score, plot_confusion_matrix
from sklearn.metrics import confusion_matrix, plot_roc_curve, precision_score
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.decomposition import PCA

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import VotingClassifier, BaggingClassifier, RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier, StackingRegressor, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB, ComplementNB


pd.set_option('display.max_columns', None)

In [3]:
df = pd.read_sas('../../data/nsch_2020_topical.sas7bdat')
mask = {
    1.0: 1.0,
    2.0: 0.0
}
df['Target'] = df['K2Q35A'].map(mask)
df.dropna(subset=['Target'], inplace=True)

y = df['Target']
X = df.drop(columns=['Target', 'K2Q35A', 'K2Q35A_1_YEARS', 'K2Q35B', 'K2Q35C', 'K2Q35D', 'AUTISMMED', 'AUTISMTREAT'])

drop_cols = []
for col in X.columns:
    null_count = X[col].isna().sum()
    null_perc = X[col].isna().sum() / len(X)
    dtype = df[col].dtype
    if null_perc > 0.5:
        drop_cols.append(col)
    if dtype == 'object':
        drop_cols.append(col)

X = X.drop(columns=drop_cols)

In [2]:
!ls ..

PhilNotebooks


In [4]:
drop_cols

['FIPSST',
 'STRATUM',
 'HHID',
 'FORMTYPE',
 'K6Q41R_STILL',
 'K6Q42R_NEVER',
 'K6Q43R_NEVER',
 'K6Q13A',
 'K6Q13B',
 'K6Q14A',
 'K6Q14B',
 'K4Q28X01',
 'K4Q28X02',
 'K4Q28X03',
 'K4Q28X_EAR',
 'K4Q28X04',
 'K4Q28X05',
 'SESPLANYR',
 'SESPLANMO',
 'K4Q37',
 'SPCSERVMO',
 'LIVEUSA_YR',
 'LIVEUSA_MO',
 'A1_LIVEUSA',
 'A2_LIVEUSA',
 'HANDS',
 'COORDINATION',
 'ERRANDALONE',
 'ALLERGIES_CURR',
 'ARTHRITIS_CURR',
 'K2Q40B',
 'K2Q61B',
 'K2Q41B',
 'K2Q42B',
 'HEART_BORN',
 'HEART_CURR',
 'HEADACHE_CURR',
 'K2Q38B',
 'K2Q33B',
 'K2Q32B',
 'SICKLECELL',
 'THALASSEMIA',
 'BLOOD_OTHER',
 'BLOOD_SCREEN',
 'CYSTFIB_SCREEN',
 'GENETIC_SCREEN',
 'K2Q34B',
 'K2Q36B',
 'K2Q60B',
 'K2Q37B',
 'K2Q30B',
 'K2Q31B',
 'K2Q31D',
 'ADDTREAT',
 'SEEKCARE',
 'CONFIRMINJURY',
 'K6Q40',
 'DOCPRIVATE',
 'K6Q10',
 'K6Q12',
 'NOTELIG',
 'AVAILABLE',
 'APPOINTMENT',
 'TRANSPORTCC',
 'NOTOPEN',
 'ISSUECOST',
 'SESCURRSVC',
 'K4Q38',
 'K5Q21',
 'TREATCHILD',
 'TREATADULT',
 'MEDHISTORY',
 'WRITEPLAN',
 'RECEIVECOPY',


In [5]:
y.value_counts()

0.0    41343
1.0     1268
Name: Target, dtype: int64

1268 autistic children is likely enough to undersample instead of oversampling; which is ideal since we have many binary columns in our dataframe.

In [6]:
X.shape

(42611, 255)

In [7]:
for col in df.columns:
    null_count = df[col].isna().sum()
    null_perc = df[col].isna().sum() / len(df)
    print(f'{col} null count: {null_count}')
    print(f'{col} null percent: {null_perc}\n')

FIPSST null count: 0
FIPSST null percent: 0.0

STRATUM null count: 0
STRATUM null percent: 0.0

HHID null count: 0
HHID null percent: 0.0

FORMTYPE null count: 0
FORMTYPE null percent: 0.0

TOTKIDS_R null count: 0
TOTKIDS_R null percent: 0.0

TENURE null count: 0
TENURE null percent: 0.0

HHLANGUAGE null count: 151
HHLANGUAGE null percent: 0.0035436859027011803

SC_AGE_YEARS null count: 0
SC_AGE_YEARS null percent: 0.0

SC_SEX null count: 0
SC_SEX null percent: 0.0

K2Q35A_1_YEARS null count: 41453
K2Q35A_1_YEARS null percent: 0.9728239187064373

BIRTH_MO null count: 260
BIRTH_MO null percent: 0.0061017108258430925

BIRTH_YR null count: 379
BIRTH_YR null percent: 0.008894416934594354

MOMAGE null count: 907
MOMAGE null percent: 0.02128558353476802

K6Q41R_STILL null count: 32554
K6Q41R_STILL null percent: 0.763981131632677

K6Q42R_NEVER null count: 30586
K6Q42R_NEVER null percent: 0.717795874304757

K6Q43R_NEVER null count: 30633
K6Q43R_NEVER null percent: 0.718898875877121

K6Q13A nul

APPOINTMENT null count: 41156
APPOINTMENT null percent: 0.9658538874938396

TRANSPORTCC null count: 41156
TRANSPORTCC null percent: 0.9658538874938396

NOTOPEN null count: 41156
NOTOPEN null percent: 0.9658538874938396

ISSUECOST null count: 41156
ISSUECOST null percent: 0.9658538874938396

HOSPITALSTAY null count: 131
HOSPITALSTAY null percent: 0.003074323531482481

K6Q15 null count: 181
K6Q15 null percent: 0.004247729459529229

SESCURRSVC null count: 36282
SESCURRSVC null percent: 0.8514702776278426

K4Q36 null count: 493
K4Q36 null percent: 0.01156978245054094

K4Q38 null count: 35448
K4Q38 null percent: 0.8318978667480228

K5Q10 null count: 225
K5Q10 null percent: 0.005280326676210368

DECISIONS null count: 6940
DECISIONS null percent: 0.16286874281288868

K5Q21 null count: 22642
K5Q21 null percent: 0.5313651404566896

TREATCHILD null count: 25434
TREATCHILD null percent: 0.59688812747882

TREATADULT null count: 32464
TREATADULT null percent: 0.7618690009621929

MEDHISTORY null cou

BULLY null count: 12670
BULLY null percent: 0.2973410621670461

RECOGBEGIN null count: 35994
RECOGBEGIN null percent: 0.8447114594822933

CLEAREXP null count: 36003
CLEAREXP null percent: 0.8449226725493417

WRITENAME null count: 36009
WRITENAME null percent: 0.8450634812607073

RECSHAPES null count: 36001
RECSHAPES null percent: 0.8448757363122198

DISTRACTED null count: 35998
DISTRACTED null percent: 0.8448053319565371

WORKTOFIN null count: 35991
WORKTOFIN null percent: 0.8446410551266105

SIMPLEINST null count: 36003
SIMPLEINST null percent: 0.8449226725493417

PLAYWELL null count: 36006
PLAYWELL null percent: 0.8449930769050246

NEWACTIVITY null count: 36002
NEWACTIVITY null percent: 0.8448992044307808

HURTSAD null count: 35999
HURTSAD null percent: 0.844828800075098

CALMDOWN null count: 36011
CALMDOWN null percent: 0.8451104174978292

TEMPER null count: 36005
TEMPER null percent: 0.8449696087864635

SITSTILL null count: 36005
SITSTILL null percent: 0.8449696087864635

RECOGABC 

Target null count: 0
Target null percent: 0.0



In [8]:
X.head()

TOTKIDS_R  TENURE  HHLANGUAGE  SC_AGE_YEARS  SC_SEX  BIRTH_MO  BIRTH_YR  \
0        2.0     1.0         1.0           3.0     1.0       8.0    2017.0   
1        1.0     1.0         1.0          14.0     2.0       2.0    2006.0   
2        1.0     1.0         1.0           1.0     2.0      10.0    2018.0   
3        2.0     1.0         1.0          15.0     2.0      10.0    2004.0   
4        2.0     2.0         1.0          16.0     2.0       8.0    2004.0   

   MOMAGE  K4Q32X01  K4Q32X02  K4Q32X03  K4Q32X04  K4Q32X05  DENTALSERV1  \
0    26.0       NaN       NaN       NaN       NaN       NaN          NaN   
1    31.0       NaN       NaN       NaN       NaN       NaN          1.0   
2    28.0       NaN       NaN       NaN       NaN       NaN          NaN   
3    29.0       1.0       2.0       2.0       2.0       2.0          1.0   
4    24.0       1.0       2.0       2.0       2.0       2.0          NaN   

   DENTALSERV2  DENTALSERV3  DENTALSERV4  DENTALSERV5  DENTALSERV6  \
0          NaN          NaN          NaN          NaN          NaN   
1          1.0          1.0          1.0          1.0          2.0   
2          NaN          NaN          NaN          NaN          NaN   
3          1.0          2.0          1.0          2.0          2.0   
4          NaN          NaN          NaN          NaN          NaN   

   DENTALSERV7  K11Q43R  A1_AGE  A2_AGE  HHCOUNT  FAMCOUNT  BREATHING  \
0          NaN      1.0    29.0    33.0      4.0       4.0        2.0   
1          2.0      1.0    43.0     NaN      3.0       2.0        2.0   
2          NaN      0.0    30.0    43.0      3.0       3.0        2.0   
3          2.0      1.0    44.0    43.0      4.0       4.0        2.0   
4          NaN      2.0    42.0    42.0      4.0       4.0        2.0   

   SWALLOWING  STOMACH  PHYSICALPAIN  TOOTHACHES  GUMBLEED  CAVITIES  \
0         2.0      2.0           2.0         2.0       2.0       2.0   
1         2.0      2.0           2.0         2.0       2.0       2.0   
2         2.0      2.0           2.0         2.0       2.0       2.0   
3         2.0      1.0           2.0         1.0       2.0       1.0   
4         2.0      NaN           2.0         2.0       2.0       2.0   

   MEMORYCOND  WALKSTAIRS  DRESSING  K2Q43B  BLINDNESS  ALLERGIES  ARTHRITIS  \
0         NaN         NaN       NaN     2.0        2.0        2.0        2.0   
1         2.0         2.0       2.0     2.0        2.0        2.0        2.0   
2         NaN         NaN       NaN     2.0        2.0        2.0        2.0   
3         2.0         2.0       2.0     2.0        2.0        2.0        2.0   
4         2.0         2.0       2.0     2.0        2.0        2.0        2.0   

   K2Q40A  K2Q61A  K2Q41A  K2Q42A  HEART  HEADACHE  K2Q38A  K2Q33A  K2Q32A  \
0     2.0     2.0     2.0     2.0    2.0       2.0     2.0     2.0     2.0   
1     1.0     2.0     2.0     2.0    2.0       2.0     2.0     1.0     2.0   
2     2.0     2.0     2.0     2.0    2.0       2.0     2.0     2.0     2.0   
3     2.0     2.0     2.0     2.0    2.0       2.0     2.0     2.0     2.0   
4     2.0     2.0     2.0     2.0    2.0       2.0     2.0     1.0     2.0   

   DOWNSYN  BLOOD  CYSTFIB  GENETIC  K2Q34A  K2Q36A  K2Q60A  K2Q37A  K2Q30A  \
0      2.0    2.0      2.0      2.0     2.0     2.0     2.0     2.0     2.0   
1      2.0    2.0      2.0      2.0     2.0     2.0     2.0     2.0     2.0   
2      2.0    2.0      2.0      2.0     2.0     2.0     2.0     2.0     2.0   
3      2.0    2.0      2.0      2.0     2.0     2.0     2.0     2.0     2.0   
4      2.0    2.0      2.0      2.0     2.0     2.0     2.0     2.0     2.0   

   K2Q31A  CONCUSSION  K2Q05  S4Q01  OVERWEIGHT  K4Q01  USUALGO  USUALSICK  \
0     2.0         2.0    1.0    1.0         2.0    1.0      1.0        1.0   
1     2.0         2.0    2.0    1.0         2.0    1.0      1.0        1.0   
2     2.0         2.0    2.0    1.0         2.0    1.0      1.0        1.0   
3     2.0         2.0    2.0    2.0         2.0

In [9]:
X.shape

(42611, 255)

In [10]:
df.describe()

TOTKIDS_R        TENURE    HHLANGUAGE  SC_AGE_YEARS        SC_SEX  \
count  42611.000000  42611.000000  42460.000000  42611.000000  42611.000000   
mean       1.840088      1.494943      1.109538      9.429631      1.483162   
std        0.862317      0.819294      0.411782      5.193992      0.499722   
min        1.000000      1.000000      1.000000      0.000000      1.000000   
25%        1.000000      1.000000      1.000000      5.000000      1.000000   
50%        2.000000      1.000000      1.000000     10.000000      1.000000   
75%        2.000000      2.000000      1.000000     14.000000      2.000000   
max        4.000000      4.000000      3.000000     17.000000      2.000000   

       K2Q35A_1_YEARS      BIRTH_MO      BIRTH_YR        MOMAGE  K6Q41R_STILL  \
count     1158.000000  42351.000000  42232.000000  41704.000000  10057.000000   
mean         5.068221      6.570258   2010.321155     30.229450      1.871930   
std          3.323733      3.409227      5.232125      5.782677      0.334184   
min          1.000000      1.000000   2002.000000     18.000000      1.000000   
25%          3.000000      4.000000   2006.000000     26.000000      2.000000   
50%          4.000000      7.000000   2010.000000     30.000000      2.000000   
75%          7.000000      9.000000   2015.000000     34.000000      2.000000   
max         15.000000     12.000000   2020.000000     45.000000      2.000000   

       K6Q42R_NEVER  K6Q43R_NEVER      K6Q13A      K6Q13B       K6Q14A  \
count  12025.000000  11978.000000  895.000000  895.000000  3519.000000   
mean       1.725821      1.923610    1.046927    1.055866     1.104291   
std        0.446118      0.265632    0.211602    0.229791     0.305681   
min        1.000000      1.000000    1.000000    1.000000     1.000000   
25%        1.000000      2.000000    1.000000    1.000000     1.000000   
50%        2.000000      2.000000    1.000000    1.000000     1.000000   
75%        2.000000      2.000000    1.000000    1.000000     1.000000   
max        2.000000      2.000000    2.000000    2.000000     2.000000   

            K6Q14B      K4Q32X01      K4Q32X02      K4Q32X03      K4Q32X04  \
count  3519.000000  26685.000000  26685.000000  26685.000000  26685.000000   
mean      1.120205      1.454075      1.569833      1.966685      1.832378   
std       0.325247      0.497896      0.495109      0.179460      0.373538   
min       1.000000      1.000000      1.000000      1.000000      1.000000   
25%       1.000000      1.000000      1.000000      2.000000      2.000000   
50%       1.000000      1.000000      2.000000      2.000000      2.000000   
75%       1.000000      2.000000      2.000000      2.000000      2.000000   
max       2.000000      2.000000      2.000000      2.000000      2.000000   

           K4Q32X05   DENTALSERV1   DENTALSERV2   DENTALSERV3   DENTALSERV4  \
count  26685.000000  33067.000000  33067.000000  33067.000000  33067.000000   
mean       1.989882      1.071945      1.083588      1.448000      1.425681   
std        0.100080      0.258400      0.276773      0.497296      0.494453   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        2.000000      1.000000      1.000000      1.000000      1.000000   
50%        2.000000      1.000000      1.000000      1.000000      1.000000   
75%        2.000000      1.000000      1.000000      2.000000      2.000000   
max        2.000000      2.000000      2.000000      2.000000      2.000000   

        DENTALSERV5   DENTALSERV6   DENTALSERV7     K4Q28X01     K4Q28X02  \
count  33067.000000  33067.000000  33067.000000  1451.000000  1451.000000   
mean       1.387516      1.803339      1.993226     1.716747     1.463818   
std        0.487191      0.397480      0.082027     0.450734     0.498861   
min        1.000000      1.000000      1.000000     1.000000     1.000000   
25%        1.000000      2.000000      2.000000     1.000000     1.000000   
50%        1.000000     